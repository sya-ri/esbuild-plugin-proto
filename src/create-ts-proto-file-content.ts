import protobuf from 'protobufjs';
import type { Options } from '@grpc/proto-loader';

export function createTSProtoFileContent(
  protoFileContent: string,
  options: Options
): string {
  const parsed = protobuf.parse(protoFileContent);

  return [
    `import { fromJSON } from '@grpc/proto-loader';`,
    `export default fromJSON(${JSON.stringify(
      parsed.root.toJSON(),
      undefined,
      2
    )}, ${JSON.stringify(options, undefined, 2)});`,
  ].join('\n');
}
