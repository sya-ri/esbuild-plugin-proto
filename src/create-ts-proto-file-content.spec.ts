import { createTSProtoFileContent } from './create-ts-proto-file-content';
import type { Options } from '@grpc/proto-loader';

describe('#createTSProtoFileContent', () => {
  test('should throw error given content is not valid proto', async () => {
    // arrange
    const protoFileContentStub = `not valid proto`;

    const optionsStub: Options = {};

    // act
    // assert
    expect(() =>
      createTSProtoFileContent(protoFileContentStub, optionsStub)
    ).toThrowErrorMatchingSnapshot();
  });

  test('should throw error given unknown syntax is passed', async () => {
    // arrange
    const protoFileContentStub = `syntax = "unknown";`;

    const optionsStub: Options = {};

    // act
    // assert
    expect(() =>
      createTSProtoFileContent(protoFileContentStub, optionsStub)
    ).toThrowErrorMatchingSnapshot();
  });

  test('should generate result given empty proto', async () => {
    // arrange
    const protoFileContentStub = ``;

    const optionsStub: Options = {};

    // act
    const result = createTSProtoFileContent(protoFileContentStub, optionsStub);

    // assert
    expect(result).toMatchSnapshot();
  });

  test('should generate result given valid proto passed with options', async () => {
    // arrange
    const protoFileContentStub = `
      syntax = "proto3";

      message Test0Response {
        Test0Details details = 1;
      }

      message Test0Details {
        int32 id = 1;
        string field0 = 2;
      }
    `;

    const optionsStub: Options = {
      arrays: true,
    };

    // act
    const result = createTSProtoFileContent(protoFileContentStub, optionsStub);

    // assert
    expect(result).toMatchSnapshot();
  });

  test('should generate result given no options passed', async () => {
    // arrange
    const protoFileContentStub = `
      syntax = "proto3";

      message Test0Response {
        Test0Details details = 1;
      }

      message Test0Details {
        int32 id = 1;
        string field0 = 2;
      }
    `;

    // act
    const result = createTSProtoFileContent(protoFileContentStub);

    // assert
    expect(result).toMatchSnapshot();
  });
});
