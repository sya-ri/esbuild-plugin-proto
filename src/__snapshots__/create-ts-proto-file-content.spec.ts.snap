// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#createTSProtoFileContent should generate result given empty proto 1`] = `
"import { fromJSON } from '@grpc/proto-loader';
export default fromJSON({}, {});"
`;

exports[`#createTSProtoFileContent should generate result given no options passed 1`] = `
"import { fromJSON } from '@grpc/proto-loader';
export default fromJSON({
  \\"nested\\": {
    \\"Test0Response\\": {
      \\"fields\\": {
        \\"details\\": {
          \\"type\\": \\"Test0Details\\",
          \\"id\\": 1
        }
      }
    },
    \\"Test0Details\\": {
      \\"fields\\": {
        \\"id\\": {
          \\"type\\": \\"int32\\",
          \\"id\\": 1
        },
        \\"field0\\": {
          \\"type\\": \\"string\\",
          \\"id\\": 2
        }
      }
    }
  }
}, undefined);"
`;

exports[`#createTSProtoFileContent should generate result given valid proto passed with options 1`] = `
"import { fromJSON } from '@grpc/proto-loader';
export default fromJSON({
  \\"nested\\": {
    \\"Test0Response\\": {
      \\"fields\\": {
        \\"details\\": {
          \\"type\\": \\"Test0Details\\",
          \\"id\\": 1
        }
      }
    },
    \\"Test0Details\\": {
      \\"fields\\": {
        \\"id\\": {
          \\"type\\": \\"int32\\",
          \\"id\\": 1
        },
        \\"field0\\": {
          \\"type\\": \\"string\\",
          \\"id\\": 2
        }
      }
    }
  }
}, {
  \\"arrays\\": true
});"
`;

exports[`#createTSProtoFileContent should throw error given content is not valid proto 1`] = `"illegal token 'not' (line 1)"`;

exports[`#createTSProtoFileContent should throw error given unknown syntax is passed 1`] = `"illegal syntax 'unknown' (line 1)"`;
